go/internal/tinder/driver_multi.go

```diff
@@ -48,126 +84,189 @@ func (md *MultiDriver) Advertise(ctx context.Context, ns string, opts ...p2p_dis
-			select {
-			case <-time.After(wait):
-			case <-ctx.Done():
-				return
-			}
+		}(i)
+	}
+	errWg.Wait()
+	for _, v := range errReturn {
+		if v == nil {
+			// At least one was successful
+			go func() {
+				wg.Wait()
+				cancelSearch()
+				close(outPeers)
+			}()
+			return outPeers, nil
 		}
-	}()
-}
-
-// FindPeers for MultiDriver doesn't care about duplicate peers, his only
-// job here is to dispatch FindPeers request across all the drivers.
-func (md *MultiDriver) FindPeers(ctx context.Context, ns string, opts ...p2p_discovery.Option) (<-chan p2p_peer.AddrInfo, error) {
-	ctx, cancel := context.WithCancel(ctx)
-
-	md.logger.Debug("looking for peers", zap.String("key", ns))
-
-	// @NOTE(gfanton): I prefer the use of select to limit the number of goroutines
-	const selDone = 0
-	selCases := make([]reflect.SelectCase, 1)
-	selCases[selDone] = reflect.SelectCase{
-		Dir:  reflect.SelectRecv,
-		Chan: reflect.ValueOf(ctx.Done()),
 	}
 
-	driverRefs := make([]string, 1)
-	for _, driver := range md.drivers {
-		ch, err := driver.FindPeers(ctx, ns, opts...)
-		if err != nil { // @TODO(gfanton): log this
-			md.logger.Warn("failed to run find peers",
-				zap.String("driver", driver.Name()),
-				zap.String("key", ns),
-				zap.Error(err))
-
-			continue
-		}
+	cancelSearch()
+	return nil, multierr.Combine(errReturn...)
+}
 
-		driverRefs = append(driverRefs, driver.Name())
-		selCases = append(selCases, reflect.SelectCase{
-			Dir:  reflect.SelectRecv,
-			Chan: reflect.ValueOf(ch),
-		})
+// FindPeersAsync expect the caller to use it asynchronously and cancel the context with enough peers or found (or deadline exceeded).
+// If the context is not canceled the search will go indefinitely.
+func (md *MultiDriver) FindPeersAsync(ctx context.Context, outPeers chan<- p2p_peer.AddrInfo, ns string, opts ...p2p_discovery.Option) error {
+	var options p2p_discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return err
 	}
-
-	ndrivers := len(selCases) - 1 // we dont want to wait for the context
-	if ndrivers == 0 {
-		md.logger.Error("no drivers available to find peers")
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
 	}
 
-	cpeers := make(chan p2p_peer.AddrInfo, ndrivers)
-	go func() {
-		defer cancel()
-		defer close(cpeers)
-
-		for ndrivers > 0 {
-			idx, value, ok := reflect.Select(selCases)
-
-			// context has been cancel stop and close chan
-			if idx == selDone {
-				md.logger.Debug("find peers done", zap.Error(ctx.Err()))
-				return
-			}
-
-			if !ok {
-				// The chosen channel has been closed, so zero out the channel to disable the case
-				selCases[idx].Chan = reflect.ValueOf(nil)
-				ndrivers--
-				continue
-			}
-
-			// we can safly get our peer
-			peer := value.Interface().(p2p_peer.AddrInfo)
-			// fmt.Printf("[multi] found a peers: %v for %s\n", peer, ns)
+	// Buffer to try to limit context switch while returning peers.
+	inPeers := make(chan p2p_peer.AddrInfo, 10)
```

Comment:
same above, 10 should be `Limit`.
And it seem that the maxLimit is quite big. Should use a defaultLimit (10 for example) in case of limit is out of range?

Reply:
> And it seem that the maxLimit is quite big. Should use a defaultLimit (10 for example) in case of limit is out of range?

I already ceil the limit if it is out of range :
```go
	limit := options.Limit
	if limit == 0 || limit > maxLimit {
		limit = maxLimit
	}
```